src/System.Data.Common/src/System/Data/EnumerableRowCollection.GetLinqDataView.cs

```diff
@@ -0,0 +1,160 @@
+    {
+        /// <summary>
+        /// Evaluates filter and sort if necessary and returns
+        /// a LinqDataView representing the LINQ query this class has collected.
+        /// </summary>
+        /// <returns>LinqDataView repesenting the LINQ query</returns>
+        internal LinqDataView GetLinqDataView() //Called by AsLinqDataView
+        {
+            if ((null == _table) || !typeof(DataRow).IsAssignableFrom(typeof(TRow)))
+            {
+                throw DataSetUtil.NotSupported(SR.ToLDVUnsupported);
+            }
+
+            LinqDataView view = null;
+
+            #region BuildSinglePredicate
+
+            Func<DataRow, bool> finalPredicate = null; //Conjunction of all .Where(..) predicates
+            if ((null != _selector) && (0 < _listOfPredicates.Count))
+            {
+                // Hook up all individual predicates into one predicate
+                // This delegate is a conjunction of multiple predicates set by the user
+                // Note: This is a Short-Circuit Conjunction
+                finalPredicate =
+                    delegate(DataRow row)
+                    {
+                        if (!Object.ReferenceEquals(row, _selector((TRow)(object)row)))
+                        {
+                            throw DataSetUtil.NotSupported(SR.ToLDVUnsupported);
+                        }
+                        foreach (Func<TRow, bool> pred in _listOfPredicates)
+                        {
+                            if (!pred((TRow)(object)row))
+                            {
+                                return false;
+                            }
+                        }
+                        return true;
+                    };
+            }
+            else if (null != _selector)
+            {
+                finalPredicate =
+                    delegate(DataRow row)
+                    {
+                        if (!Object.ReferenceEquals(row, _selector((TRow)(object)row)))
+                        {
+                            throw DataSetUtil.NotSupported(SR.ToLDVUnsupported);
+                        }
+                        return true;
+                    };
+            }
+            else if (0 < _listOfPredicates.Count)
+            {
+                finalPredicate =
+                    delegate(DataRow row)
+                    {
+                        foreach (Func<TRow, bool> pred in _listOfPredicates)
+                        {
+                            if (!pred((TRow)(object)row))
+                            {
+                                return false;
+                            }
+                        }
+                        return true;
+                    };
+            }
+            #endregion BuildSinglePredicate
+
+            #region Evaluate Filter/Sort
+            //  All of this mess below is because we want to create index only once.
```

Comment:
is there a better word than mess? 😸 complexity sounds a little bit more .. intentional